File: velocity/example-plugin/src/main/java/com/example/ExamplePlugin.java
package com.example;

import com.google.inject.Inject;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
import com.velocitypowered.api.event.proxy.ProxyShutdownEvent;
import com.velocitypowered.api.plugin.Plugin;
import dev.faststats.core.ErrorTracker;
import dev.faststats.core.Metrics;
import dev.faststats.core.Settings;
import dev.faststats.core.data.Metric;
import dev.faststats.velocity.VelocityMetrics;
import org.jspecify.annotations.Nullable;

@Plugin(id = "example", name = "Example Plugin", version = "1.0.0",
        url = "https://example.com", authors = {"Your Name"})
public class ExamplePlugin {
    private final VelocityMetrics.Factory metricsFactory;
    private @Nullable Metrics metrics = null;

    @Inject
    public ExamplePlugin(final VelocityMetrics.Factory factory) {
        this.metricsFactory = factory;
    }

    @Subscribe
    public void onProxyInitialize(final ProxyInitializeEvent event) {
        this.metrics = metricsFactory
                // Custom metrics require a corresponding data source in your project settings
                .addMetric(Metric.number("example_metric", () -> 42))

                // Error tracking must be enabled in the project settings
                .errorTracker(ErrorTracker.contextAware())

                .settings(Settings.withToken("YOUR_TOKEN_HERE")) // token can be found in the settings of your project
                .create(this);
    }

    @Subscribe
    public void onProxyStop(final ProxyShutdownEvent event) {
        if (metrics != null) metrics.shutdown(); // safely shut down metrics submission
    }
}

File: velocity/example-plugin/src/main/java/com/example/ExamplePlugin.kt
package com.example

import com.google.inject.Inject
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent
import com.velocitypowered.api.event.proxy.ProxyShutdownEvent
import com.velocitypowered.api.plugin.Plugin
import dev.faststats.ErrorTracker
import dev.faststats.data.Metric
import dev.faststats.velocity.VelocityContext
import java.util.concurrent.atomic.AtomicInteger

@Plugin(
    id = "example",
    name = "Example Plugin",
    version = "1.0.0",
    url = "https://example.com",
    authors = ["Your Name"],
)
class ExamplePlugin @Inject constructor(contextBuilder: VelocityContext.Builder) {
    private val gameCount = AtomicInteger()

    private val context = contextBuilder
        .token("YOUR_TOKEN_HERE")
        .errorTrackerService(ERROR_TRACKER)
        // .metrics(Metrics.Factory::create) // Define a minimal metrics instance without any custom metrics
        .metrics { factory ->
            factory
                // Custom metrics require a corresponding data source in your project settings
                .addMetric(Metric.number("game_count") { gameCount.get() })
                .addMetric(Metric.string("server_version") { "1.0.0" })

                // #onFlush is invoked after successful metrics submission
                // This is useful for cleaning up cached data
                .onFlush { gameCount.set(0) } // reset game count on flush

                .create()
        }
        .create()

    @Subscribe
    fun onProxyInitialize(event: ProxyInitializeEvent) {
        context.ready() // start metrics and errors submission
    }

    @Subscribe
    fun onProxyStop(event: ProxyShutdownEvent) {
        context.shutdown() // safely shut down configured services
    }

    fun startGame() {
        gameCount.incrementAndGet()
    }

    companion object {
        val ERROR_TRACKER: ErrorTracker = ErrorTracker.contextAware()
    }
}
